src/main/java/net/dv8tion/jda/bot/entities/impl/DefaultShardManagerImpl.java

```diff
@@ -0,0 +1,336 @@
+import net.dv8tion.jda.core.entities.Game;
+import net.dv8tion.jda.core.entities.impl.JDAImpl;
+import net.dv8tion.jda.core.exceptions.RateLimitedException;
+import net.dv8tion.jda.core.hooks.IEventManager;
+import net.dv8tion.jda.core.managers.impl.PresenceImpl;
+import net.dv8tion.jda.core.requests.SessionReconnectQueue;
+import net.dv8tion.jda.core.utils.Checks;
+import net.dv8tion.jda.core.utils.SimpleLog;
+import okhttp3.OkHttpClient;
+
+/**
+ * JDA's default {@link net.dv8tion.jda.bot.sharding.ShardManager ShardManager} implementation.
+ * To  create new instances use the {@link net.dv8tion.jda.bot.sharding.DefaultShardManagerBuilder DefaultShardManagerBuilder}.
+ *
+ * @since  3.4
+ * @author Aljoscha Grebe
+ */
+public class DefaultShardManagerImpl implements ShardManager
+{
+    public static final SimpleLog LOG = SimpleLog.getLog(ShardManager.class);
+
+    protected final IAudioSendFactory audioSendFactory;
+    protected final boolean autoReconnect;
+    protected final int corePoolSize;
+    protected final boolean enableBulkDeleteSplitting;
+    protected final boolean enableVoice;
+    protected final IEventManager eventManager;
+    protected final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(r ->
+    {
+        final Thread t = new Thread(r, "DefaultShardManagerImpl");
+        t.setDaemon(true);
+        t.setPriority(Thread.NORM_PRIORITY + 1);
+        return t;
+    });
+    protected final Game game;
+    protected final boolean idle;
+    protected final List<Object> listeners;
+    protected final int maxReconnectDelay;
+    protected final Queue<Integer> queue = new ConcurrentLinkedQueue<>();
+    protected ShardCacheViewImpl shards = new ShardCacheViewImpl();
+    protected int shardsTotal;
+    protected final int backoff;
+    protected final OkHttpClient.Builder httpClientBuilder;
+    protected final WebSocketFactory wsFactory;
+    protected final SessionReconnectQueue reconnectQueue;
+    protected final ShardedRateLimiter shardedRateLimiter;
+    protected final AtomicBoolean shutdown = new AtomicBoolean(false);
+    protected final Thread shutdownHook;
+    protected final OnlineStatus status;
+    protected final String token;
+    protected ScheduledFuture<?> worker;
+
+    public DefaultShardManagerImpl(final int shardsTotal, final Collection<Integer> shardIds, final List<Object> listeners, final String token,
+            final IEventManager eventManager, final IAudioSendFactory audioSendFactory, final Game game, final OnlineStatus status,
+            final OkHttpClient.Builder httpClientBuilder, final WebSocketFactory wsFactory, ShardedRateLimiter shardedRateLimiter,
+            final int maxReconnectDelay, final int corePoolSize, final boolean enableVoice, final boolean enableShutdownHook,
+            final boolean enableBulkDeleteSplitting, final boolean autoReconnect, final boolean idle, final SessionReconnectQueue reconnectQueue,
+            final int backoff)
```

Comment:
Maybe write documentation on these fields? I see this being a public class that can be extended.